sopel/plugin.py

```diff
@@ -820,31 +823,168 @@ def add_attribute(function):
 
     Rate-limited functions that use scheduled future commands should import
     :class:`threading.Timer` instead of :mod:`sched`, or rate limiting will
     not work properly.
+
+    .. versionchanged:: 8.0
+
+        Optional keyword argument ``message`` was added in Sopel 8.
+
+    .. note::
+
+        Users on the admin list in Sopel’s configuration are exempted from rate
+        limits.
+
+    .. seealso::
+
+        You can control each rate limit separatly, with their own custom
+        message using :func:`rate_user`, :func:`rate_channel`, or
+        :func:`rate_server`.
+
+    """
+    def add_attribute(function):
+        if not hasattr(function, 'rate'):
+            function.rate = user
+        if not hasattr(function, 'channel_rate'):
+            function.channel_rate = channel
+        if not hasattr(function, 'global_rate'):
+            function.global_rate = server
+        function.default_rate_message = message
+        return function
+    return add_attribute
+
+
+def rate_user(
+    rate: int,
+    message: typing.Optional[str] = None,
+) -> typing.Callable:
+    """Decorate a function to be rate-limited for a user.
+
+    :param rate: seconds between permitted calls of this function by the same
+                 user
+    :param message: optional; message send as notice when a user hits the limit
+
+    This decorator can be used alone or with the :func:`rate` decorator, as it
+    will always take precedence::
+
+        @rate(10, 10, 10)
+        @rate_user(20, 'You hit your rate limit for this function.')
+            # user limit will be set to 20, other to 10
+            # will send a NOTICE only when a user hits their own limit
+            # as other rate limit don't have any message set
+
+    If you don't provide a message, the default message set (if any) by
+    :func:`rate` will be used instead.
+
+    .. versionadded:: 8.0
+    """
+    def add_attribute(function):
+        function.rate = rate
+        function.user_rate_message = message
+        return function
+    return add_attribute
+
+
+def rate_channel(
+    rate: int,
+    message: typing.Optional[str] = None,
+) -> typing.Callable:
+    """Decorate a function to be rate-limited for a channel.
+
+    :param rate: seconds between permitted calls of this function in the same
+                 channel, regardless of triggering user
+    :param message: optional; message send as notice when a user hits the limit
+
+    This decorator can be used alone or with the :func:`rate` decorator, as it
+    will always take precedence::
+
+        @rate(10, 10, 10)
+        @rate_channel(5, 'You hit the channel rate limit for this function.')
+            # channel limit will be set to 5, other to 10
+            # will send a NOTICE only when a user hits the channel limit
+            # as other rate limit don't have any message set
+
+    If you don't provide a message, the default message set (if any) by
+    :func:`rate` will be used instead.
```

Comment:
I have not applied this to every instance, as I'm not 100% sure the line-wrapping aligns with the standard, so I'm leaving it to you to tweak this if necessary. 😸

```suggestion
    If you don't provide a message, the default message set by :func:`rate`
    (if any) will be used instead.
```